File: plugin/java/src/main/java/io/github/jinganix/webpb/java/utils/Imports.java
/*
 * Copyright (c) 2020 The Webpb Authors, All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * https://github.com/jinganix/webpb
 */

package io.github.jinganix.webpb.java.utils;

import static io.github.jinganix.webpb.utilities.utils.OptionUtils.getOpts;

import com.github.javaparser.JavaParser;
import com.google.protobuf.Descriptors.FileDescriptor;
import com.google.protobuf.Descriptors.GenericDescriptor;
import io.github.jinganix.webpb.utilities.descriptor.WebpbExtend;
import io.github.jinganix.webpb.utilities.utils.Const;
import io.github.jinganix.webpb.utilities.utils.DescriptorUtils;
import io.github.jinganix.webpb.utilities.utils.OptionUtils;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.apache.commons.lang3.StringUtils;

/** Utility to handle imports. */
public class Imports {

  private static final JavaParser JAVA_PARSER = new JavaParser();

  private final List<ImportPath> imported = new ArrayList<>();

  private final String javaPackage;

  private final List<ImportPath> lookup;

  private final ImportMapper importMapper;

  /**
   * Constructor.
   *
   * @param javaPackage java package
   * @param lookup list of {@link ImportPath}
   */
  public Imports(String javaPackage, List<ImportPath> lookup) {
    this.javaPackage = javaPackage;
    this.lookup = lookup;
    this.importMapper = new ImportMapper();
  }

  /**
   * Get lookup paths.
   *
   * @param fd {@link FileDescriptor}
   * @return list of {@link ImportPath}
   */
  public static List<ImportPath> getLookup(FileDescriptor fd) {
    return Stream.of(
            Arrays.asList(
                "java.lang.Integer",
                "java.lang.Long",
                "java.lang.String",
                "java.util.List",
                "java.util.Map",
                Const.RUNTIME_PACKAGE + ".WebpbMessage",
                Const.RUNTIME_PACKAGE + ".WebpbMeta",
                Const.RUNTIME_PACKAGE + ".WebpbSubValue",
                Const.RUNTIME_PACKAGE + ".common.InQuery",
                Const.RUNTIME_PACKAGE + ".enumeration.Enumeration"),
            OptionUtils.getWebpbOpts(fd, WebpbExtend.FileOpts::hasJava).getJava().getImportList(),
            getOpts(fd, WebpbExtend.FileOpts::hasJava).getJava().getImportList(),
            DescriptorUtils.resolveTopLevelTypes(fd).stream()
                .map(
                    e ->
                        StringUtils.isEmpty(GeneratorUtils.getJavaPackage(e))
                            ? e.getName()
                            : GeneratorUtils.getJavaPackage(e) + "." + e.getName())
                .collect(Collectors.toList()))
        .flatMap(List::stream)
        .map(ImportPath::new)
        .sorted(((o1, o2) -> Integer.compare(o2.getPath().length(), o1.getPath().length())))
        .distinct()
        .collect(Collectors.toList());
  }

  /**
   * Import an annotation.
   *
   * @param str annotation, eg: @Anno("abc")
   * @return imported annotation
   */
  public String importAnnotation(String str) {
    return JAVA_PARSER
        .parseAnnotation(str)
        .getResult()
        .map(
            expr -> {
              expr.accept(new ImportVisitor(), this);
              return expr.toString();
            })
        .orElseThrow(() -> new RuntimeException("Bad annotation: " + str));
  }

  /**
   * Import a class or interface.
   *
   * @param str a class or interface, eg: List&lt;String&gt;
   * @return imported class or interface
   */
  public String importClassOrInterface(String str) {
    return JAVA_PARSER
        .parseClassOrInterfaceType(str)
        .getResult()
        .map(
            expr -> {
              expr.accept(new ImportVisitor(), this);
              return expr.toString();
            })
        .orElseThrow(() -> new RuntimeException("Bad class or interface: " + str));
  }

  /**
   * Import from an enum or a message descriptor.
   *
   * @param descriptor {@link GenericDescriptor}
   * @return imported string
   */
  public String importGenericDescriptor(GenericDescriptor descriptor) {
    FileDescriptor fd = descriptor.getFile();
    String packageName = fd.getPackage();
    String relative = descriptor.getFullName().replace(packageName + ".", "");
    String prefix =
        StringUtils.isEmpty(GeneratorUtils.getJavaPackage(fd))
            ? ""
            : GeneratorUtils.getJavaPackage(fd) + ".";
    String fullPath = prefix + relative;
    String importPath = prefix + relative.split("\\.")[0];
    return checkAndImport(fullPath, importPath, relative);
  }

  /**
   * Find qualified name in imported paths.
   *
   * @param name qualified or simple name
   * @return qualified name or itself
   */
  public String importedQualifiedName(String name) {
    for (ImportPath importPath : imported) {
      String relative = importPath.relative(name);
      if (relative != null) {
        return importPath.getPath();
      }
    }
    return name;
  }

  /**
   * Import a name (eg. foo.bar) or simple name (eg. foo).
   *
   * @param name a name or simple name
   * @return imported name eg. name a.b.c -> return c
   */
  public String importName(String name) {
    for (ImportPath importPath : lookup) {
      String relative = importPath.relative(name);
      if (relative != null) {
        return checkAndImport(name, importPath.getPath(), relative);
      }
    }
    throw new RuntimeException("No import path found for: " + name);
  }

  private String checkAndImport(String fullPath, String importPath, String relative) {
    String mapped = importMapper.map(importPath);
    int index = mapped.lastIndexOf(".");
    String importPackage = index < 0 ? mapped : mapped.substring(0, index);
    if (importPackage.equals(this.javaPackage) || importPackage.equals(importPath)) {
      return relative;
    }
    String identifier = relative.split("\\.")[0];
    for (ImportPath imported : imported) {
      if (!imported.getPath().equals(mapped) && imported.getIdentifier().equals(identifier)) {
        return importMapper.map(fullPath);
      }
    }
    if (imported.stream().noneMatch(e -> e.getPath().equals(mapped))) {
      imported.add(new ImportPath(mapped));
    }
    return relative;
  }

  /**
   * To imported path strings.
   *
   * @return imported paths
   */
  public List<String> toList() {
    return imported.stream()
        .map(e -> importMapper.map(e.getPath()))
        .filter(StringUtils::isNotEmpty)
        .distinct()
        .sorted(StringUtils::compare)
        .collect(Collectors.toList());
  }
}

File: plugin/internal/javagen/imports.go
package javagen

import (
	"fmt"
	"sort"
	"strings"

	"github.com/jinganix/webpb/plugin/internal/core"
	"google.golang.org/protobuf/reflect/protoreflect"
)

// Imports manages Java import statements.
type Imports struct {
	javaPackage string
	imported    []ImportPath
	lookup      []ImportPath
	mapper      *ImportMapper
	rootFd      protoreflect.FileDescriptor
}

// NewImports creates an Imports instance.
func NewImports(javaPackage string, lookup []ImportPath, rootFd protoreflect.FileDescriptor) *Imports {
	return &Imports{
		javaPackage: javaPackage,
		lookup:      lookup,
		mapper:      NewImportMapper(),
		rootFd:      rootFd,
	}
}

// GetLookup builds lookup paths for a file descriptor.
func GetLookup(fd protoreflect.FileDescriptor) ([]ImportPath, error) {
	if fd == nil {
		var lookup []ImportPath
		for _, path := range []string{
			"java.lang.Boolean",
			"java.lang.Double",
			"java.lang.Float",
			"java.lang.Integer",
			"java.lang.Long",
			"java.lang.String",
			"java.util.List",
			"java.util.Map",
			core.RuntimePackage + ".WebpbMessage",
			core.RuntimePackage + ".WebpbMeta",
			core.RuntimePackage + ".WebpbSubValue",
			core.RuntimePackage + ".common.InQuery",
			core.RuntimePackage + ".enumeration.Enumeration",
		} {
			importPath, err := NewImportPath(path)
			if err != nil {
				return nil, err
			}
			lookup = append(lookup, importPath)
		}
		return lookup, nil
	}
	var paths []string
	paths = append(paths,
		"java.lang.Boolean",
		"java.lang.Double",
		"java.lang.Float",
		"java.lang.Integer",
		"java.lang.Long",
		"java.lang.String",
		"java.util.List",
		"java.util.Map",
		core.RuntimePackage+".WebpbMessage",
		core.RuntimePackage+".WebpbMeta",
		core.RuntimePackage+".WebpbSubValue",
		core.RuntimePackage+".common.InQuery",
		core.RuntimePackage+".enumeration.Enumeration",
	)
	paths = append(paths, core.GetWebpbFileOpts(fd, core.HasFileJava).GetJava().GetImport()...)
	paths = append(paths, core.GetFileOpts(fd, core.HasFileJava).GetJava().GetImport()...)
	for _, d := range core.ResolveTopLevelTypes(fd) {
		pkg := GetJavaPackage(d)
		if pkg == "" {
			paths = append(paths, string(d.Name()))
		} else {
			paths = append(paths, pkg+"."+string(d.Name()))
		}
	}
	seen := map[string]struct{}{}
	var lookup []ImportPath
	for _, path := range paths {
		if _, ok := seen[path]; ok {
			continue
		}
		seen[path] = struct{}{}
		importPath, err := NewImportPath(path)
		if err != nil {
			return nil, err
		}
		lookup = append(lookup, importPath)
	}
	sort.Slice(lookup, func(i, j int) bool {
		return len(lookup[i].path) > len(lookup[j].path)
	})
	return lookup, nil
}

// ImportAnnotation imports types referenced in an annotation string.
func (i *Imports) ImportAnnotation(str string) (string, error) {
	parser := &annotationParser{imports: i}
	return parser.parseAnnotation(str)
}

// ImportClassOrInterface imports types referenced in a class or interface type string.
func (i *Imports) ImportClassOrInterface(str string) (string, error) {
	parser := &typeParser{imports: i}
	return parser.parseClassOrInterfaceType(str)
}

// ImportGenericDescriptor imports a message or enum descriptor type.
func (i *Imports) ImportGenericDescriptor(descriptor protoreflect.Descriptor) (string, error) {
	if descriptor == nil {
		return "", fmt.Errorf("Bad class or interface: <nil>")
	}
	fd := descriptor.ParentFile()
	if fd == nil {
		fd = core.ResolveDescriptorFile(i.rootFd, descriptor)
	}
	if fd == nil {
		return i.ImportClassOrInterface(string(descriptor.Name()))
	}
	packageName := string(fd.Package())
	relative := strings.Replace(string(descriptor.FullName()), packageName+".", "", 1)
	relative = strings.TrimPrefix(relative, ".")
	prefix := ""
	if pkg := GetJavaPackage(fd); pkg != "" {
		prefix = pkg + "."
	}
	fullPath := prefix + relative
	importPath := prefix + strings.Split(relative, ".")[0]
	return i.checkAndImport(fullPath, importPath, relative), nil
}

// ImportedQualifiedName returns the qualified name if imported.
func (i *Imports) ImportedQualifiedName(name string) string {
	for _, importPath := range i.imported {
		if rel := importPath.Relative(name); rel != "" {
			return importPath.Path()
		}
	}
	return name
}

func (i *Imports) importName(name string) (string, error) {
	for _, importPath := range i.lookup {
		if rel, ok := importPath.relative(name); ok {
			return i.checkAndImport(name, importPath.Path(), rel), nil
		}
	}
	return "", fmt.Errorf("No import path found for: %s", name)
}

func (i *Imports) checkAndImport(fullPath, importPath, relative string) string {
	mapped := i.mapper.Map(importPath)
	importPackage := mapped
	if idx := strings.LastIndex(mapped, "."); idx >= 0 {
		importPackage = mapped[:idx]
	}
	if importPackage == i.javaPackage || importPackage == importPath {
		return relative
	}
	identifier := strings.Split(relative, ".")[0]
	for _, imported := range i.imported {
		if imported.Path() != mapped && imported.Identifier() == identifier {
			return i.mapper.Map(fullPath)
		}
	}
	if !i.containsImport(mapped) {
		i.imported = append(i.imported, ImportPath{path: mapped, identifier: identifierFromPath(mapped)})
	}
	return relative
}

func identifierFromPath(path string) string {
	if i := strings.LastIndex(path, "."); i >= 0 {
		return path[i+1:]
	}
	return path
}

func (i *Imports) containsImport(path string) bool {
	for _, imported := range i.imported {
		if imported.Path() == path {
			return true
		}
	}
	return false
}

// ToList returns sorted import statements.
func (i *Imports) ToList() []string {
	seen := map[string]struct{}{}
	var out []string
	for _, e := range i.imported {
		mapped := i.mapper.Map(e.Path())
		if mapped == "" {
			continue
		}
		if _, ok := seen[mapped]; ok {
			continue
		}
		seen[mapped] = struct{}{}
		out = append(out, mapped)
	}
	sort.Strings(out)
	return out
}
